File: src/main/java/org/athenian/PairsCount.java
package org.athenian;

import java.util.stream.IntStream;

public class PairsCount {

    public static int pairsCountLoop(String s) {
        int cnt = 0;
        for (int i = 0; i < s.length() - 1; i++)
            if (s.substring(i, i + 1).equals(s.substring(i + 1, i + 2)))
                cnt++;
        return cnt;
    }

    public static int pairsCountRecursion(String s) {
        if (s.length() <= 1)
            return 0;

        String firstLetter = s.substring(0, 1);
        String secondLetter = s.substring(1, 2);
        String remainingLetters = s.substring(1);

        if (firstLetter.equals(secondLetter))
            return 1 + pairsCountRecursion(remainingLetters);
        else
            return pairsCountRecursion(remainingLetters);
    }

    public static int pairsCountStream(String s) {
        return (int) IntStream
                .range(0, s.length() - 1)
                .filter(i -> s.substring(i, i + 1).equals(s.substring(i + 1, i + 2)))
                .mapToLong(value -> 1)
                .sum();
    }
}

File: src/main/kotlin/org/athenian/PairsCount.kt
package org.athenian

import java.util.stream.IntStream


fun pairsCountLoop(s: String): Int {
    var cnt = 0
    repeat(s.length - 1) {
        if (s[it] == s[it + 1])
            cnt++
    }
    return cnt
}

fun pairsCountRecursion(s: String): Int =
    when {
        s.length <= 1 -> 0
        else -> (if (s[0] == s[1]) 1 else 0) + pairsCountRecursion(s.substring(1))
    }

fun pairsCountStream(s: String): Int {
    return IntStream
        .range(0, s.length - 1)
        .filter { i -> s[i] == s[i + 1] }
        .count()
        .toInt()
}
